### ISSUE TITLE
Linking fail with lots of undefined references probably due to: {ar,ranlib} plugin needed to handle lto object

### ISSUE THREAD

**Issue by viniciusferrao** [no-fee-ethereum-mining/nsfminer#36]
**Describe the bug**
When trying to compile the software on RHEL 8.3 with GCC 8.3.1 and CUDA 11.2 linking fails with a lot of undefined references and there's some non stopping error messages regarding a missing lot plugin during link phase of libraries when running ar or ranlib.

```
[  0%] Built target nsfminer-buildinfo-git
Scanning dependencies of target nsfminer-buildinfo
[  3%] Building C object CMakeFiles/nsfminer-buildinfo.dir/nsfminer/buildinfo.c.o
[  6%] Linking C static library nsfminer/libnsfminer-buildinfo.a
/usr/bin/ar: CMakeFiles/nsfminer-buildinfo.dir/nsfminer/buildinfo.c.o: plugin needed to handle lto object
/usr/bin/ranlib: nsfminer/libnsfminer-buildinfo.a(buildinfo.c.o): plugin needed to handle lto object
[ 10%] Built target nsfminer-buildinfo
[ 26%] Built target devcore
[ 43%] Built target hwmon
[ 46%] Linking CXX static library libethash-cuda.a
/usr/bin/ar: CMakeFiles/ethash-cuda.dir/CUDAMiner.cpp.o: plugin needed to handle lto object
/usr/bin/ranlib: libethash-cuda.a(CUDAMiner.cpp.o): plugin needed to handle lto object
[ 53%] Built target ethash-cuda
[ 66%] Built target ethcore
Scanning dependencies of target poolprotocols
[ 70%] Building CXX object libpoolprotocols/CMakeFiles/poolprotocols.dir/stratum/EthStratumClient.cpp.o
[ 73%] Linking CXX static library libpoolprotocols.a
/usr/bin/ar: CMakeFiles/poolprotocols.dir/PoolURI.cpp.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/poolprotocols.dir/PoolManager.cpp.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/poolprotocols.dir/testing/SimulateClient.cpp.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/poolprotocols.dir/stratum/EthStratumClient.cpp.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/poolprotocols.dir/getwork/EthGetworkClient.cpp.o: plugin needed to handle lto object
/usr/bin/ranlib: libpoolprotocols.a(PoolURI.cpp.o): plugin needed to handle lto object
/usr/bin/ranlib: libpoolprotocols.a(PoolManager.cpp.o): plugin needed to handle lto object
/usr/bin/ranlib: libpoolprotocols.a(SimulateClient.cpp.o): plugin needed to handle lto object
/usr/bin/ranlib: libpoolprotocols.a(EthStratumClient.cpp.o): plugin needed to handle lto object
/usr/bin/ranlib: libpoolprotocols.a(EthGetworkClient.cpp.o): plugin needed to handle lto object
[ 86%] Built target poolprotocols
Scanning dependencies of target apicore
[ 90%] Building CXX object libapicore/CMakeFiles/apicore.dir/ApiServer.cpp.o
[ 93%] Linking CXX static library libapicore.a
/usr/bin/ar: CMakeFiles/apicore.dir/ApiServer.cpp.o: plugin needed to handle lto object
/usr/bin/ranlib: libapicore.a(ApiServer.cpp.o): plugin needed to handle lto object
[ 93%] Built target apicore
Scanning dependencies of target nsfminer
[ 96%] Building CXX object nsfminer/CMakeFiles/nsfminer.dir/main.cpp.o

[100%] Linking CXX executable nsfminer
/tmp/ccoZfEA0.ltrans0.ltrans.o: In function `dev::eth::TelemetryType::strvec(std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) [clone .constprop.63]':
<artificial>:(.text+0x6956): undefined reference to `g_logOptions'
/tmp/ccoZfEA0.ltrans0.ltrans.o: In function `headers(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, bool) [clone .constprop.34]':
<artificial>:(.text+0x98f4): undefined reference to `nsfminer_get_buildinfo'
/tmp/ccoZfEA0.ltrans2.ltrans.o: In function `dev::LogOutputStream<dev::NoteChannel>::~LogOutputStream() [clone .lto_priv.391]':
<artificial>:(.text+0x9e68): undefined reference to `dev::simpleDebugOut(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/tmp/ccoZfEA0.ltrans2.ltrans.o: In function `main':
<artificial>:(.text.startup+0x22): undefined reference to `dev::setThreadName(char const*)'
<artificial>:(.text.startup+0x2c1): undefined reference to `dev::s_fixedHashEngine'
<artificial>:(.text.startup+0x59a): undefined reference to `dev::setenv(char const*, char const*, bool)'
<artificial>:(.text.startup+0x5ab): undefined reference to `dev::setenv(char const*, char const*, bool)'
<artificial>:(.text.startup+0x5bc): undefined reference to `dev::setenv(char const*, char const*, bool)'
<artificial>:(.text.startup+0x5cd): undefined reference to `dev::setenv(char const*, char const*, bool)'
<artificial>:(.text.startup+0x17aa): undefined reference to `g_logSyslog'
<artificial>:(.text.startup+0x17b3): undefined reference to `g_logNoColor'
<artificial>:(.text.startup+0x17c8): undefined reference to `g_logNoColor'
<artificial>:(.text.startup+0x1821): undefined reference to `dev::NoteChannel::severity()'
<artificial>:(.text.startup+0x1830): undefined reference to `dev::LogOutputStreamBase::LogOutputStreamBase(int)'
<artificial>:(.text.startup+0x1dbf): undefined reference to `g_logOptions'
<artificial>:(.text.startup+0x1e16): undefined reference to `g_logNoColor'
<artificial>:(.text.startup+0x1e69): undefined reference to `g_logSyslog'
<artificial>:(.text.startup+0x2954): undefined reference to `dev::URI::URI(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool)'
<artificial>:(.text.startup+0x2e87): undefined reference to `dev::URI::KnownSchemes[abi:cxx11](dev::ProtocolFamily)'
<artificial>:(.text.startup+0x3edb): undefined reference to `dev::URI::URI(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool)'
/tmp/ccoZfEA0.ltrans3.ltrans.o: In function `MinerCLI::cliDisplayInterval_elapsed(boost::system::error_code const&)':
<artificial>:(.text+0x4184): undefined reference to `dev::NoteChannel::severity()'
<artificial>:(.text+0x418e): undefined reference to `dev::LogOutputStreamBase::LogOutputStreamBase(int)'
<artificial>:(.text+0x419f): undefined reference to `dev::eth::Farm::m_this'
<artificial>:(.text+0x4358): undefined reference to `dev::eth::Farm::m_this'
<artificial>:(.text+0x43f1): undefined reference to `dev::NoteChannel::severity()'
<artificial>:(.text+0x4400): undefined reference to `dev::LogOutputStreamBase::LogOutputStreamBase(int)'
/tmp/ccoZfEA0.ltrans3.ltrans.o: In function `MinerCLI::signalHandler(int)':
<artificial>:(.text+0x45b3): undefined reference to `dev::setThreadName(char const*)'
<artificial>:(.text+0x45cf): undefined reference to `dev::CritChannel::severity()'
<artificial>:(.text+0x45d9): undefined reference to `dev::LogOutputStreamBase::LogOutputStreamBase(int)'
<artificial>:(.text+0x463a): undefined reference to `dev::setThreadName(char const*)'
/tmp/ccoZfEA0.ltrans3.ltrans.o: In function `MinerCLI::doMiner()':
<artificial>:(.text+0x4f5a): undefined reference to `dev::eth::PoolManager::PoolManager(dev::eth::PoolSettings)'
<artificial>:(.text+0x4fc0): undefined reference to `dev::NoteChannel::severity()'
<artificial>:(.text+0x4fcf): undefined reference to `dev::LogOutputStreamBase::LogOutputStreamBase(int)'
<artificial>:(.text+0x51d9): undefined reference to `ApiServer::ApiServer(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
<artificial>:(.text+0x521d): undefined reference to `dev::eth::PoolManager::m_this'
<artificial>:(.text+0x5222): undefined reference to `dev::eth::PoolManager::start()'
<artificial>:(.text+0x5308): undefined reference to `ApiServer::stop()'
<artificial>:(.text+0x530f): undefined reference to `dev::eth::PoolManager::m_this'
<artificial>:(.text+0x5320): undefined reference to `dev::NoteChannel::severity()'
<artificial>:(.text+0x532f): undefined reference to `dev::LogOutputStreamBase::LogOutputStreamBase(int)'
<artificial>:(.text+0x5430): undefined reference to `ApiServer::start()'
<artificial>:(.text+0x543c): undefined reference to `dev::eth::PoolManager::m_this'
<artificial>:(.text+0x5441): undefined reference to `dev::eth::PoolManager::stop()'
/tmp/ccoZfEA0.ltrans3.ltrans.o: In function `MinerCLI::execute()':
<artificial>:(.text+0x564b): undefined reference to `dev::eth::Farm::Farm(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, dev::eth::DeviceDescriptor, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, dev::eth::DeviceDescriptor> > >&, dev::eth::FarmSettings)'
<artificial>:(.text+0x5a3a): undefined reference to `dev::getFormattedMemory[abi:cxx11](double, dev::ScaleSuffix, int)'
<artificial>:(.text+0x5b53): undefined reference to `dev::eth::CUDAMiner::enumDevices(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, dev::eth::DeviceDescriptor, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, dev::eth::DeviceDescriptor> > >&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [nsfminer/CMakeFiles/nsfminer.dir/build.make:128: nsfminer/nsfminer] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:468: nsfminer/CMakeFiles/nsfminer.dir/all] Error 2
gmake: *** [Makefile:171: all] Error 2
```

**To Reproduce**
1. Compile the software
2. Check the errors during CMake build phase

**Expected behavior**
Able to properly link the software.

**Environment (please complete the following information):**
- Operating System: Red Hat Enterprise Linux 8.3
- Hardware: Dell Poweredge C4140 (2x Xeon Gold 6130 + 4x nVIDIA Tesla V100)
- Ethminer Version: Master
- Ethminer options used: 
```
-- ETHASHCL         Build OpenCL components                      OFF
-- ETHASHCUDA       Build CUDA components                        ON
-- ETHASHCPU        Build CPU components (only for development)  OFF
-- ETHDBUS          Build D-Bus components                       OFF
-- APICORE          Build API Server components                  ON
-- BINKERN          Install AMD binary kernels                   OFF
-- DEVBUILD         Build with dev logging                       OFF
```

**Additional context**
Package versions:

```
[root@rhvc4140 build]# rpm -qa | egrep "gcc|cuda"
gcc-8.3.1-5.1.el8.x86_64
cuda-cuxxfilt-11-2-11.2.67-1.x86_64
cuda-cudart-devel-11-2-11.2.72-1.x86_64
cuda-gdb-11-2-11.2.67-1.x86_64
nvidia-driver-cuda-libs-460.32.03-1.el8.x86_64
cuda-libraries-devel-11-2-11.2.0-1.x86_64
cuda-nvvp-11-2-11.2.67-1.x86_64
cuda-nsight-compute-11-2-11.2.0-1.x86_64
cuda-cuobjdump-11-2-11.2.67-1.x86_64
cuda-nvrtc-11-2-11.2.67-1.x86_64
cuda-memcheck-11-2-11.2.67-1.x86_64
cuda-nvml-devel-11-2-11.2.67-1.x86_64
cuda-sanitizer-11-2-11.2.67-1.x86_64
cuda-cupti-11-2-11.2.67-1.x86_64
cuda-toolkit-11-2-11.2.0-1.x86_64
cuda-cudart-11-2-11.2.72-1.x86_64
cuda-nvtx-11-2-11.2.67-1.x86_64
cuda-command-line-tools-11-2-11.2.0-1.x86_64
cuda-11-2-11.2.0-1.x86_64
cuda-driver-devel-11-2-11.2.72-1.x86_64
cuda-nvrtc-devel-11-2-11.2.67-1.x86_64
cuda-runtime-11-2-11.2.0-1.x86_64
cuda-demo-suite-11-2-11.2.67-1.x86_64
cuda-visual-tools-11-2-11.2.0-1.x86_64
cuda-drivers-460.32.03-1.x86_64
gcc-c++-8.3.1-5.1.el8.x86_64
cuda-nvdisasm-11-2-11.2.67-1.x86_64
cuda-nsight-systems-11-2-11.2.0-1.x86_64
cuda-tools-11-2-11.2.0-1.x86_64
cuda-nsight-11-2-11.2.67-1.x86_64
nvidia-driver-cuda-460.32.03-1.el8.x86_64
cuda-nvprof-11-2-11.2.67-1.x86_64
cuda-documentation-11-2-11.2.67-1.x86_64
cuda-nvcc-11-2-11.2.67-1.x86_64
cuda-nvprune-11-2-11.2.67-1.x86_64
cuda-11.2.0-1.x86_64
libgcc-8.3.1-5.1.el8.x86_64
cuda-libraries-11-2-11.2.0-1.x86_64
cuda-compiler-11-2-11.2.0-1.x86_64
cuda-samples-11-2-11.2.67-1.x86_64
```

**Comment by viniciusferrao** [comment]
Alright, I've a solution.

This is an issue that I was only able to replicate on Enterprise Linux systems (RHEL and CentOS). For whatever reasons `ranlib` and `ar` fails to handle the LTO objects. But the system ships with wrappers around both `ranlib` and `ar`; they are: `gcc-ranlib` and `gcc-ar`.

During the compile phase if we pass the options `-DCMAKE_AR=/usr/bin/gcc-ar -DCMAKE_RANLIB=/usr/bin/gcc-ranlib` to `cmake` we can properly (almost) properly build the software on EL systems.

I've looked in Debian / Ubuntu systems and passing those arguments, with correct paths, don't break the software, so it appears to be safe to apply this fix system wide.

**Comment by jean-m-cyr** [comment]
Wouldn't it just be easier to kill the -flto option? My tests don't show ant measurable difference with plain -O3

**Comment by jean-m-cyr** [comment]
Fixed by PR #44

**Comment by viniciusferrao** [comment]
@jean-m-cyr I understand that compiler optimizations may not be relevant in a software that the power came from the GPU. Although is it a good idea do remove the feature/optimizations?

I agree that the easy solution is just wipeout -flto. But there's a fix that I've proposed with `-DCMAKE_AR=/usr/bin/gcc-ar -DCMAKE_RANLIB=/usr/bin/gcc-ranlib`. I could make a PR for this, I just need to learn a little about `cmake` because I'm too much old school and usually I create plain `Makefile`s by hand.

There isn't an easy way to setup *AR* and *RANLIB* directly on `CMakeLists.txt`? So it can be figured out during `cmake ..` phase setting *AR* and *RANLIB* to the `gcc` counterparts?
